test/integration/api/user-preferences.js

```diff
@@ -0,0 +1,127 @@
+const { testService } = require('../setup');
+
+describe('api: user-preferences', () => {
+
+  it('should not allow storing preferences for nonexistent projects', testService(async (service) => {
+    const asAlice = await service.login('alice');
+
+    await asAlice.put('/v1/user-preferences/project/123/someProperty')
+      .send({ propertyValue: 'someValue' })
+      .expect(400); // project 123 doesn't exist
+  }));
+
+  it('PUT/DELETE preference storage operations', testService(async (service) => {
```

Comment:
I'd like to see more/smaller tests! 

In this big test, the main themes i'm getting out of it are:
- "it should store site-wide and project-specific preferences of varying complexity"
- "it should not return preferences that do not belong to the user"
- "it should group project preferences by project"

I think you should keep that test, or maybe split it up into 2 or 3 tests, but maybe there could also be stupidly basic tests like:
- "it allows a site-wide preference to be set"
- "it allows a project-specific preference to be set"
- "it allows a preference to be deleted"
- "it returns the (non blank) preferences in the extended metadata for the current user"

And one of our original concerns that the design of the api alleviates: 
- "it allows a preference to be modified without overwriting another preference""

I'd also be interested in tests of the failure cases
- no propertyValue supplied
- empty string value supplied
- (anything else that tests the database constraints or other error handling in the resources)

Reply:
Yeah this is pretty much the situation for all Central tests. 

> ...just leads down the path of incrementally building up a whole bunch of DB state in one transaction. Unless I'm missing something that's just the nature of this thing. If I want to test the aggregation I need to do a whole bunch of PUTs....

Send a bunch of requests about forms/submissions to build up the state. If it gets super repetitive, we make special setup functions for specific batches of tests, but it's still making the requests for every test.

Reply:
Thanks for adding some of the smaller tests! e.g.  
- validates the request body stringently 
- can store a JS null propertyValue
- should not allow storing preferences for nonexistent projects

Reply:
I can still see an advantage of smaller tests so i can just look at an individual test and grok the 1 thing its trying to do instead of having to follow the whole journey, but eh, it's fine for now!